core/modules/loader/ServerTcpBase.cc

```diff
@@ -0,0 +1,618 @@
+            boost::asio::transfer_at_least(bytesInMsg),
+            boost::bind(
+                    &TcpBaseConnection::_handleShiftFromRight1,
+                    shared_from_this(),
+                    boost::asio::placeholders::error,
+                    boost::asio::placeholders::bytes_transferred
+            )
+    );
+}
+
+
+void TcpBaseConnection::_handleShiftFromRight1(boost::system::error_code const& ec, size_t bytesTrans) {
+    std::string const funcName = "_handleShiftFromRight1";
+    if (ec) {
+        LOGS(_log, LOG_LVL_ERROR, funcName << " ec=" << ec);
+        _freeConnect();
+        return;
+    }
+    // Fix the buffer with the information given.
+    _buf.advanceWriteCursor(bytesTrans);
+    LOGS(_log, LOG_LVL_INFO, funcName << " bytes=" << bytesTrans << " _buf" << _buf.dumpStr(false));
+    try {
+        // TODO move as much of this to CentralWorker as possible
+        LOGS(_log, LOG_LVL_INFO, funcName << " parsing bytes=" << bytesTrans << " _buf" << _buf.dumpStr(false));
+        auto protoKeyShiftReq = StringElement::protoParse<proto::KeyShiftRequest>(_buf);
+        if (protoKeyShiftReq == nullptr) {
+            throw LoaderMsgErr(ERR_LOC, " KeyShiftRequest parse failure ");
+        }
+        // Extract keysToShift from the protobuffer
+        int keyShiftReq = protoKeyShiftReq->keystoshift();
+        if (keyShiftReq < 1) {
+            throw LoaderMsgErr(ERR_LOC, " KeyShiftRequest for < 1 key");
+        }
+        // Build and send the KeyList message back (send smallest keys to right node)
+        StringElement::UPtr keyList = _serverTcpBase->getCentralWorker()->buildKeyList(keyShiftReq);
+        auto keyListTransmitSz = keyList->transmitSize();
+        BufferUdp data(keyListTransmitSz);
+        keyList->appendToData(data);
+        ServerTcpBase::writeData(_socket, data);
+
+
+        // Wait for the SHIFT_FROM_RIGHT_KEYS_RECEIVED response back.
+        _buf.reset();
+        auto msgElem = _buf.readFromSocket(_socket, funcName +" waiting for SHIFT_FROM_RIGHT_KEYS_RECEIVED");
+        UInt32Element::Ptr received = std::dynamic_pointer_cast<UInt32Element>(msgElem);
+        if (received == nullptr || received->element !=  LoaderMsg::SHIFT_FROM_RIGHT_RECEIVED) {
+            LOGS(_log, LOG_LVL_INFO, funcName << " did not get SHIFT_FROM_RIGHT_RECEIVED");
+            throw LoaderMsgErr(ERR_LOC, " receive failure");
+        }
+        _serverTcpBase->getCentralWorker()->finishShiftFromRight();
+        LOGS(_log, LOG_LVL_INFO, funcName << " done dumpKeys " <<
+                                 _serverTcpBase->getCentralWorker()->dumpKeysStr(2));
+    } catch (LoaderMsgErr &msgErr) {
+        LOGS(_log, LOG_LVL_ERROR, msgErr.what());
+        LOGS(_log, LOG_LVL_ERROR, funcName << " key shift failed");
+        _freeConnect();
+        return;
+    }
+    boost::system::error_code ecode;
+    _readKind(ecode, 0); // get next message
```

Comment:
I've already made a comment earlier on calling this method explicitly. Consider refactoring `_readKind`